src/openzaak/utils/fields.py

```diff
@@ -33,19 +50,86 @@ def set_attributes_from_name(self, name):
         super().set_attributes_from_name(name)
         self.column = self.source_field.column
 
-    def contribute_to_class(self, cls, name, private_only=False):
-        super().contribute_to_class(cls, name, private_only=True)
-        setattr(cls, name, self)
 
-    def __get__(self, instance, instance_type=None):
-        return getattr(instance, self.source_field.name)
+class AliasServiceUrlField(AliasMixin, ServiceUrlField):
+    def __init__(
+        self,
+        source_field: ServiceUrlField,
+        allow_write_when: Callable = lambda i: True,
+        **kwargs,
+    ):
+        self.source_field = source_field
+        self.allow_write_when = allow_write_when
 
-    def __set__(self, instance, value):
-        if not self.allow_write_when(instance):
-            return
-        setattr(instance, self.source_field.name, value)
+        super().__init__(
+            base_field=source_field.base_field,
+            relative_field=source_field.relative_field,
+            **kwargs,
+        )
 
 
 # register the override, as the upstream RelativeDeltaField registers its own admin
 # form field override as well.
 FORMFIELD_FOR_DBFIELD_DEFAULTS[DurationField] = {"form_class": RelativeDeltaFormField}
+
+
+class FkOrServiceUrlField(FkOrURLField):
+    """ supports zgw_comnsumers.ServiceUrlField as 'url_field' """
+
+    def _add_check_constraint(
+        self, options, name="{prefix}{fk_field}_or_{url_base_field}_filled"
+    ) -> None:
+        """
+        Create the DB constraints and add them if they're not present yet.
+        """
+        if self.null:
+            return
+
+        # during migrations, the FK fields are added later, causing the constraint SQL
+        # building to blow up. We can ignore this at that time.
+        if self.model.__module__ == "__fake__":
+            return
+
+        url_base_field = self._url_field.base_field
+        # one of both MUST be filled and they cannot be filled both at the
+        # same time
+        empty_url_base_field = models.Q(**{f"{url_base_field}__isnull": True})
+        empty_fk_field = models.Q(**{f"{self.fk_field}__isnull": True})
+        fk_filled = ~empty_fk_field & empty_url_base_field
+        url_filled = empty_fk_field & ~empty_url_base_field
```

Comment:
I think this needs the addition that the relative part is not empty too

Reply:
I've added the separate constraint for `ServiceUrlField` (PR - https://github.com/maykinmedia/zgw-consumers/pull/56)